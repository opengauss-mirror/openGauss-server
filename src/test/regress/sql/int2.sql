
--
-- INT2
--

CREATE TABLE INT2_TBL(f1 int2);

INSERT INTO INT2_TBL(f1) VALUES ('0   ');

INSERT INTO INT2_TBL(f1) VALUES ('  1234 ');

INSERT INTO INT2_TBL(f1) VALUES ('    -1234');

INSERT INTO INT2_TBL(f1) VALUES ('34.5');

-- largest and smallest values
INSERT INTO INT2_TBL(f1) VALUES ('32767');
INSERT INTO INT2_TBL(f1) VALUES ('-32768');

-- bad input values -- should give errors
INSERT INTO INT2_TBL(f1) VALUES ('32768');
INSERT INTO INT2_TBL(f1) VALUES ('-32769');
INSERT INTO INT2_TBL(f1) VALUES ('100000');
INSERT INTO INT2_TBL(f1) VALUES ('asdf');
INSERT INTO INT2_TBL(f1) VALUES ('    ');
INSERT INTO INT2_TBL(f1) VALUES ('- 1234');
INSERT INTO INT2_TBL(f1) VALUES ('4 444');
INSERT INTO INT2_TBL(f1) VALUES ('123 dt');
INSERT INTO INT2_TBL(f1) VALUES ('');
INSERT INTO INT2_TBL(f1) VALUES (sqrt(-2));
INSERT INTO INT2_TBL SELECT f1/0 FROM INT2_TBL;

SELECT '' AS five, * FROM INT2_TBL ORDER BY f1;

SELECT '' AS four, i.* FROM INT2_TBL i WHERE i.f1 <> int2 '0' ORDER BY f1;

SELECT '' AS four, i.* FROM INT2_TBL i WHERE i.f1 <> int4 '0' ORDER BY f1;

SELECT '' AS one, i.* FROM INT2_TBL i WHERE i.f1 = int2 '0' ORDER BY f1;

SELECT '' AS one, i.* FROM INT2_TBL i WHERE i.f1 = int4 '0' ORDER BY f1;

SELECT '' AS two, i.* FROM INT2_TBL i WHERE i.f1 < int2 '0' ORDER BY f1;

SELECT '' AS two, i.* FROM INT2_TBL i WHERE i.f1 < int4 '0' ORDER BY f1;

SELECT '' AS three, i.* FROM INT2_TBL i WHERE i.f1 <= int2 '0' ORDER BY f1;

SELECT '' AS three, i.* FROM INT2_TBL i WHERE i.f1 <= int4 '0' ORDER BY f1;

SELECT '' AS two, i.* FROM INT2_TBL i WHERE i.f1 > int2 '0' ORDER BY f1;

SELECT '' AS two, i.* FROM INT2_TBL i WHERE i.f1 > int4 '0' ORDER BY f1;

SELECT '' AS three, i.* FROM INT2_TBL i WHERE i.f1 >= int2 '0' ORDER BY f1;

SELECT '' AS three, i.* FROM INT2_TBL i WHERE i.f1 >= int4 '0' ORDER BY f1;

-- positive odds 
SELECT '' AS one, i.* FROM INT2_TBL i WHERE (i.f1 % int2 '2') = int2 '1' ORDER BY f1;

-- any evens 
SELECT '' AS three, i.* FROM INT2_TBL i WHERE (i.f1 % int4 '2') = int2 '0' ORDER BY f1;

SELECT '' AS five, i.f1, i.f1 * int2 '2' AS x FROM INT2_TBL i ORDER BY f1;

SELECT '' AS five, i.f1, i.f1 * int2 '2' AS x FROM INT2_TBL i
WHERE abs(f1) < 16384 ORDER BY f1;

SELECT '' AS five, i.f1, i.f1 * int4 '2' AS x FROM INT2_TBL i ORDER BY f1;

SELECT '' AS five, i.f1, i.f1 + int2 '2' AS x FROM INT2_TBL i ORDER BY f1;

SELECT '' AS five, i.f1, i.f1 + int2 '2' AS x FROM INT2_TBL i
WHERE f1 < 32766 ORDER BY f1;

SELECT '' AS five, i.f1, i.f1 + int4 '2' AS x FROM INT2_TBL i ORDER BY f1;

SELECT '' AS five, i.f1, i.f1 - int2 '2' AS x FROM INT2_TBL i ORDER BY f1;

SELECT '' AS five, i.f1, i.f1 - int2 '2' AS x FROM INT2_TBL i
WHERE f1 > -32767 ORDER BY f1;

SELECT '' AS five, i.f1, i.f1 - int4 '2' AS x FROM INT2_TBL i ORDER BY f1;

SELECT '' AS five, i.f1, i.f1 / int2 '2' AS x FROM INT2_TBL i ORDER BY f1;

SELECT '' AS five, i.f1, i.f1 / int4 '2' AS x FROM INT2_TBL i ORDER BY f1;

-- divide zero
SELECT (0)::int2 / (-2)::int2;
SELECT (0)::int2 / (-2)::int4;
SELECT (0)::int2 / (-2)::int8;

-- corner cases
SELECT (-1::int2<<15)::text;
SELECT ((-1::int2<<15)+1::int2)::text;

-- check sane handling of INT16_MIN overflow cases
SELECT (-32768)::int2 * (-1)::int2;
SELECT (-32768)::int2 / (-1)::int2;
SELECT (-32768)::int2 % (-1)::int2;

set behavior_compat_options ='unbind_divide_bound';
SELECT (-32768)::int2 / (-1)::int2;
reset behavior_compat_options;

-- check overflow case during dn insert
CREATE TABLE t_num(a numeric(2,1));
CREATE FUNCTION test_insert(IN iter int) RETURNS integer AS $$
BEGIN
	forall i in 1..iter	
		INSERT INTO t_num SELECT f1 FROM INT2_TBL;
	return iter;
END;
$$ LANGUAGE plpgsql;
SELECT test_insert(1);
DROP FUNCtION test_insert;
DROP TABLE t_num;

