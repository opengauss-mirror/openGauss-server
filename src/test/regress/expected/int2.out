--
-- INT2
--
CREATE TABLE INT2_TBL(f1 int2);
INSERT INTO INT2_TBL(f1) VALUES ('0   ');
INSERT INTO INT2_TBL(f1) VALUES ('  1234 ');
INSERT INTO INT2_TBL(f1) VALUES ('    -1234');
INSERT INTO INT2_TBL(f1) VALUES ('34.5');
ERROR:  invalid input syntax for integer: "34.5"
LINE 1: INSERT INTO INT2_TBL(f1) VALUES ('34.5');
                                         ^
CONTEXT:  referenced column: f1
-- largest and smallest values
INSERT INTO INT2_TBL(f1) VALUES ('32767');
INSERT INTO INT2_TBL(f1) VALUES ('-32768');
-- bad input values -- should give errors
INSERT INTO INT2_TBL(f1) VALUES ('32768');
ERROR:  value "32768" is out of range for type smallint
LINE 1: INSERT INTO INT2_TBL(f1) VALUES ('32768');
                                         ^
CONTEXT:  referenced column: f1
INSERT INTO INT2_TBL(f1) VALUES ('-32769');
ERROR:  value "-32769" is out of range for type smallint
LINE 1: INSERT INTO INT2_TBL(f1) VALUES ('-32769');
                                         ^
CONTEXT:  referenced column: f1
INSERT INTO INT2_TBL(f1) VALUES ('100000');
ERROR:  value "100000" is out of range for type smallint
LINE 1: INSERT INTO INT2_TBL(f1) VALUES ('100000');
                                         ^
CONTEXT:  referenced column: f1
INSERT INTO INT2_TBL(f1) VALUES ('asdf');
ERROR:  invalid input syntax for integer: "asdf"
LINE 1: INSERT INTO INT2_TBL(f1) VALUES ('asdf');
                                         ^
CONTEXT:  referenced column: f1
INSERT INTO INT2_TBL(f1) VALUES ('    ');
ERROR:  invalid input syntax for integer: "    "
LINE 1: INSERT INTO INT2_TBL(f1) VALUES ('    ');
                                         ^
CONTEXT:  referenced column: f1
INSERT INTO INT2_TBL(f1) VALUES ('- 1234');
ERROR:  invalid input syntax for integer: "- 1234"
LINE 1: INSERT INTO INT2_TBL(f1) VALUES ('- 1234');
                                         ^
CONTEXT:  referenced column: f1
INSERT INTO INT2_TBL(f1) VALUES ('4 444');
ERROR:  invalid input syntax for integer: "4 444"
LINE 1: INSERT INTO INT2_TBL(f1) VALUES ('4 444');
                                         ^
CONTEXT:  referenced column: f1
INSERT INTO INT2_TBL(f1) VALUES ('123 dt');
ERROR:  invalid input syntax for integer: "123 dt"
LINE 1: INSERT INTO INT2_TBL(f1) VALUES ('123 dt');
                                         ^
CONTEXT:  referenced column: f1
INSERT INTO INT2_TBL(f1) VALUES ('');
INSERT INTO INT2_TBL(f1) VALUES (sqrt(-2));
ERROR:  cannot take square root of a negative number
CONTEXT:  referenced column: f1
INSERT INTO INT2_TBL SELECT f1/0 FROM INT2_TBL;
ERROR:  division by zero
CONTEXT:  referenced column: f1
SELECT '' AS five, * FROM INT2_TBL ORDER BY f1;
 five |   f1   
------+--------
      | -32768
      |  -1234
      |      0
      |   1234
      |  32767
      |       
(6 rows)

SELECT '' AS four, i.* FROM INT2_TBL i WHERE i.f1 <> int2 '0' ORDER BY f1;
 four |   f1   
------+--------
      | -32768
      |  -1234
      |   1234
      |  32767
(4 rows)

SELECT '' AS four, i.* FROM INT2_TBL i WHERE i.f1 <> int4 '0' ORDER BY f1;
 four |   f1   
------+--------
      | -32768
      |  -1234
      |   1234
      |  32767
(4 rows)

SELECT '' AS one, i.* FROM INT2_TBL i WHERE i.f1 = int2 '0' ORDER BY f1;
 one | f1 
-----+----
     |  0
(1 row)

SELECT '' AS one, i.* FROM INT2_TBL i WHERE i.f1 = int4 '0' ORDER BY f1;
 one | f1 
-----+----
     |  0
(1 row)

SELECT '' AS two, i.* FROM INT2_TBL i WHERE i.f1 < int2 '0' ORDER BY f1;
 two |   f1   
-----+--------
     | -32768
     |  -1234
(2 rows)

SELECT '' AS two, i.* FROM INT2_TBL i WHERE i.f1 < int4 '0' ORDER BY f1;
 two |   f1   
-----+--------
     | -32768
     |  -1234
(2 rows)

SELECT '' AS three, i.* FROM INT2_TBL i WHERE i.f1 <= int2 '0' ORDER BY f1;
 three |   f1   
-------+--------
       | -32768
       |  -1234
       |      0
(3 rows)

SELECT '' AS three, i.* FROM INT2_TBL i WHERE i.f1 <= int4 '0' ORDER BY f1;
 three |   f1   
-------+--------
       | -32768
       |  -1234
       |      0
(3 rows)

SELECT '' AS two, i.* FROM INT2_TBL i WHERE i.f1 > int2 '0' ORDER BY f1;
 two |  f1   
-----+-------
     |  1234
     | 32767
(2 rows)

SELECT '' AS two, i.* FROM INT2_TBL i WHERE i.f1 > int4 '0' ORDER BY f1;
 two |  f1   
-----+-------
     |  1234
     | 32767
(2 rows)

SELECT '' AS three, i.* FROM INT2_TBL i WHERE i.f1 >= int2 '0' ORDER BY f1;
 three |  f1   
-------+-------
       |     0
       |  1234
       | 32767
(3 rows)

SELECT '' AS three, i.* FROM INT2_TBL i WHERE i.f1 >= int4 '0' ORDER BY f1;
 three |  f1   
-------+-------
       |     0
       |  1234
       | 32767
(3 rows)

-- positive odds 
SELECT '' AS one, i.* FROM INT2_TBL i WHERE (i.f1 % int2 '2') = int2 '1' ORDER BY f1;
 one |  f1   
-----+-------
     | 32767
(1 row)

-- any evens 
SELECT '' AS three, i.* FROM INT2_TBL i WHERE (i.f1 % int4 '2') = int2 '0' ORDER BY f1;
 three |   f1   
-------+--------
       | -32768
       |  -1234
       |      0
       |   1234
(4 rows)

SELECT '' AS five, i.f1, i.f1 * int2 '2' AS x FROM INT2_TBL i ORDER BY f1;
ERROR:  smallint out of range
CONTEXT:  referenced column: x
SELECT '' AS five, i.f1, i.f1 * int2 '2' AS x FROM INT2_TBL i
WHERE abs(f1) < 16384 ORDER BY f1;
ERROR:  smallint out of range
SELECT '' AS five, i.f1, i.f1 * int4 '2' AS x FROM INT2_TBL i ORDER BY f1;
 five |   f1   |   x    
------+--------+--------
      | -32768 | -65536
      |  -1234 |  -2468
      |      0 |      0
      |   1234 |   2468
      |  32767 |  65534
      |        |       
(6 rows)

SELECT '' AS five, i.f1, i.f1 + int2 '2' AS x FROM INT2_TBL i ORDER BY f1;
ERROR:  smallint out of range
CONTEXT:  referenced column: x
SELECT '' AS five, i.f1, i.f1 + int2 '2' AS x FROM INT2_TBL i
WHERE f1 < 32766 ORDER BY f1;
 five |   f1   |   x    
------+--------+--------
      | -32768 | -32766
      |  -1234 |  -1232
      |      0 |      2
      |   1234 |   1236
(4 rows)

SELECT '' AS five, i.f1, i.f1 + int4 '2' AS x FROM INT2_TBL i ORDER BY f1;
 five |   f1   |   x    
------+--------+--------
      | -32768 | -32766
      |  -1234 |  -1232
      |      0 |      2
      |   1234 |   1236
      |  32767 |  32769
      |        |       
(6 rows)

SELECT '' AS five, i.f1, i.f1 - int2 '2' AS x FROM INT2_TBL i ORDER BY f1;
ERROR:  smallint out of range
CONTEXT:  referenced column: x
SELECT '' AS five, i.f1, i.f1 - int2 '2' AS x FROM INT2_TBL i
WHERE f1 > -32767 ORDER BY f1;
 five |  f1   |   x   
------+-------+-------
      | -1234 | -1236
      |     0 |    -2
      |  1234 |  1232
      | 32767 | 32765
(4 rows)

SELECT '' AS five, i.f1, i.f1 - int4 '2' AS x FROM INT2_TBL i ORDER BY f1;
 five |   f1   |   x    
------+--------+--------
      | -32768 | -32770
      |  -1234 |  -1236
      |      0 |     -2
      |   1234 |   1232
      |  32767 |  32765
      |        |       
(6 rows)

SELECT '' AS five, i.f1, i.f1 / int2 '2' AS x FROM INT2_TBL i ORDER BY f1;
 five |   f1   |    x    
------+--------+---------
      | -32768 |  -16384
      |  -1234 |    -617
      |      0 |       0
      |   1234 |     617
      |  32767 | 16383.5
      |        |        
(6 rows)

SELECT '' AS five, i.f1, i.f1 / int4 '2' AS x FROM INT2_TBL i ORDER BY f1;
 five |   f1   |    x    
------+--------+---------
      | -32768 |  -16384
      |  -1234 |    -617
      |      0 |       0
      |   1234 |     617
      |  32767 | 16383.5
      |        |        
(6 rows)

-- divide zero
SELECT (0)::int2 / (-2)::int2;
 ?column? 
----------
        0
(1 row)

SELECT (0)::int2 / (-2)::int4;
 ?column? 
----------
        0
(1 row)

SELECT (0)::int2 / (-2)::int8;
 ?column? 
----------
        0
(1 row)

-- corner cases
SELECT (-1::int2<<15)::text;
  text  
--------
 -32768
(1 row)

SELECT ((-1::int2<<15)+1::int2)::text;
  text  
--------
 -32767
(1 row)

-- check sane handling of INT16_MIN overflow cases
SELECT (-32768)::int2 * (-1)::int2;
ERROR:  smallint out of range
SELECT (-32768)::int2 / (-1)::int2;
ERROR:  smallint out of range
SELECT (-32768)::int2 % (-1)::int2;
 ?column? 
----------
        0
(1 row)

set behavior_compat_options ='unbind_divide_bound';
SELECT (-32768)::int2 / (-1)::int2;
 ?column? 
----------
    32768
(1 row)

reset behavior_compat_options;
-- check overflow case during dn insert
CREATE TABLE t_num(a numeric(2,1));
CREATE FUNCTION test_insert(IN iter int) RETURNS integer AS $$
BEGIN
	forall i in 1..iter	
		INSERT INTO t_num SELECT f1 FROM INT2_TBL;
	return iter;
END;
$$ LANGUAGE plpgsql;
SELECT test_insert(1);
ERROR:  numeric field overflow
DETAIL:  A field with precision 2, scale 1 must round to an absolute value less than 10^1.
CONTEXT:  referenced column: a
SQL statement "INSERT INTO t_num SELECT f1 FROM INT2_TBL"
PL/pgSQL function test_insert(integer) line 4 at SQL statement
referenced column: test_insert
DROP FUNCtION test_insert;
DROP TABLE t_num;
